d2/src/main/java/com/linkedin/d2/balancer/servers/ZooKeeperAnnouncer.java

```diff
@@ -826,4 +822,25 @@ private ImmutablePair<String, String> getZnodePathAndData(String cluster) {
     }
     return new ImmutablePair<>(nodePath, nodeData);
   }
+
+
+  private void validatePartitionData(Map<Integer, PartitionData> partitionData) {
+    for (Map.Entry<Integer, PartitionData> entry : partitionData.entrySet()) {
+      double weight = entry.getValue().getWeight();
+      // check max weight
+      if (_maxWeight != null && weight > _maxWeight) {
+        _log.warn("", new IllegalArgumentException(String.format("[ACTION NEEDED] Weight %f in Partition %d is greater"
+            + " than the max weight allowed: %d. Please correct the weight. It will be force-capped to the max weight "
+            + "in the future.", weight, entry.getKey(), _maxWeight)));
+        // TODO: throw exception or cap weight to max weight
+      }
+
+      // check decimal places
+      if (BigDecimal.valueOf(weight).scale() > 2) {
+        _log.warn("", new IllegalArgumentException(String.format("Weight %f in Partition %d has more than 2 decimal "
+            + "places. It will be rounded to 2 decimal places in the future.", weight, entry.getKey())));
+        // TODO: round weight to 2 decimal places.
+      }
+    }
+  }
```

Comment:
```suggestion
 private void validatePartitionData(Map<Integer, PartitionData> partitionData) {
    partitionData.forEach((partitionId, data) -> {
      final double weight = data.getWeight();
      // check max weight
      if (_maxWeight != null && weight > _maxWeight) {
        _log.warn("", new IllegalArgumentException(String.format("[ACTION NEEDED] Weight %f in Partition %d is greater"
                                                                     + " than the max weight allowed: %d. Please correct the weight. It will be force-capped to the max weight "
                                                                     + "in the future.", weight, partitionId, _maxWeight)));
        // TODO: throw exception or cap weight to max weight
      }

      // check decimal places
      if (BigDecimal.valueOf(weight).scale() > 2) {
        _log.warn("", new IllegalArgumentException(String.format("Weight %f in Partition %d has more than 2 decimal "
                                                                     + "places. It will be rounded to 2 decimal places in the future.", weight, partitionId)));
        // TODO: round weight to 2 decimal places.
      }
    });
  }

```

Reply:
Updated the code with configurable actions, and we need to modify the entry value when the bad value needs to be rectified. So we need to keep the for-loop.